### PR TITLE
expression, plan: rewrite builtin arithmetic function PLUS

### PR DESCRIPTION
**Pull request by zz-jason**
to https://github.com/pingcap/tidb/issues/3851

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
+package expression
```
Comment (shenli) [inline expression/builtin_arithmetic.go]: Miss license.

File: expression/builtin.go [hunk 2]
```diff
@@ -549,6 +549,15 @@ func (b *baseFunctionClass) verifyArgs(args []Expression) error {
 	if l < b.minArgs || (b.maxArgs != -1 && l > b.maxArgs) {
 		return errIncorrectParameterCount.GenByArgs(b.funcName)
 	}
+	for i, arg := range args {
+		if arg == nil {
```
Comment (shenli) [inline expression/builtin.go]: When will we meet this?

Comment (zz-jason) [inline expression/builtin.go]: fix ut `TestSubstituteCorCol2Constant` in file "expression/util_test.go"

In that situation, builtin function `plus`'s second argument is a `Column` with nil return type, this is invalid in the new expression architecture.

**Comment by shenli** [comment]
@zz-jason CI failed.

**Comment by zz-jason** [comment]
@shenli @XuHuaiyu @hanfei1991 PTAL

**Comment by hanfei1991** [comment]
@zz-jason CI failed :(

File: expression/builtin_arithmetic.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+		sig := &builtinArithmeticPlusDecimalSig{baseDecimalBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	}
+}
+
+type builtinArithmeticPlusIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (s *builtinArithmeticPlusIntSig) evalInt(row []types.Datum) (val int64, isNull bool, err error) {
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: should we consider the unsigned flag for plusInt?

File: expression/builtin_arithmetic.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	baseFunctionClass
+	op opcode.Op
+}
+
+func (c *arithmeticFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinArithmeticSig{newBaseBuiltinFunc(args, ctx), c.op}
+	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: no need to check c.verifyArgs(args) again.

File: expression/builtin_arithmetic.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+)
+
+type arithmeticPlusFunctionClass struct {
+	baseFunctionClass
+}
+
+// getEvalTp will get the `evalTp` of the argument according to it's return type.
+func (c *arithmeticPlusFunctionClass) getEvalTp(ft *types.FieldType) evalTp {
+	switch ft.Tp {
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong,
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: Can these be rewritten using types.ClassType?

Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: you can refer to this: https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.cc#L1194

Comment (zz-jason) [inline expression/builtin_arithmetic.go]: no, for TypeDuration, `types.ClassType` will return `ClassString`, but we need `ClassInt` here.

File: expression/builtin_arithmetic.go (hunk 3 above)
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: fixed

File: executor/executor_test.go [hunk 6]
```diff
@@ -1399,6 +1399,26 @@ func (s *testSuite) TestMathBuiltin(c *C) {
...
+	defer func() {
+		s.cleanEnv(c)
+		testleak.AfterTest(c)()
+	}()
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	// for plus
+	tk.MustExec("DROP TABLE IF EXISTS t;")
+	tk.MustExec("CREATE TABLE t(a DECIMAL(4, 2), b DECIMAL(5, 3));")
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: add column for unsigned.

File: expression/builtin_arithmetic.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+		return tpInt
+	case mysql.TypeNewDecimal, mysql.TypeDatetime:
+		return tpDecimal
+	}
+	return tpReal
+}
+
+// setFlenDecimal4Int is called to set proper `Flen` and `Decimal` of return
+// type according to the two input parameter's types.
+func (c *arithmeticPlusFunctionClass) setFlenDecimal4Int(retTp, a, b *types.FieldType) {
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: can this be refered when set flen and decimal.
```
sql/ item_func.cc
644:void Item_func::fix_num_length_and_dec()

```

**Comment by shenli** [comment]
CI failed and jekins failed.

File: executor/executor_test.go (hunk 6 above)
Comment (zz-jason) [inline executor/executor_test.go]: added

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by zz-jason** [comment]
blocked by: https://github.com/pingcap/tidb/issues/3986

File: expression/builtin_arithmetic_test.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+	for _, tc := range testCases {
+		sig, err := funcs[ast.Plus].getFunction(datumsToConstants(types.MakeDatums(tc.args...)), s.ctx)
+		c.Assert(err, IsNil)
+		c.Assert(sig, NotNil)
+		val, err := sig.eval(nil)
+		c.Assert(err, IsNil)
+		c.Assert(val, testutil.DatumEquals, types.NewDatum(tc.expect))
+	}
+}
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic_test.go]: add a test case for IsDeterminstic.

File: util/types/field_type.go [hunk 9]
```diff
@@ -1169,3 +1169,13 @@ func SetBinChsClnFlag(ft *FieldType) {
 	ft.Collate = charset.CollationBin
 	ft.Flag |= mysql.BinaryFlag
 }
+
+// IsTemporalType checks if a type is a temporal type.
+func IsTemporalType(tp byte) bool {
```
Comment (XuHuaiyu) [inline util/types/field_type.go]: move this to util/types/etc.go

File: expression/builtin_arithmetic.go [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+		retTp.Flen = int(math.Max(float64(a.Flen), float64(b.Flen))) + 1
+		retTp.Flen = int(math.Min(float64(retTp.Flen), float64(mysql.MaxIntWidth)))
+	}
+}
+
+// setFlenDecimal4Real is called to set proper `Flen` and `Decimal` of return
+// type according to the two input parameter's types.
+func (c *arithmeticPlusFunctionClass) setFlenDecimal4RealOrDecimal(retTp, a, b *types.FieldType, isReal bool) {
+	if a.Decimal != types.UnspecifiedLength && b.Decimal != types.UnspecifiedLength {
+		retTp.Decimal = int(math.Max(float64(a.Decimal), float64(b.Decimal)))
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: we can set retTp.Decimal to int(math.Max(float64(a.Decimal), float64(b.Decimal))) despite of whether a.Decimal or b.Decimal is unspecified.

``` go
retTp.Flen, retTp.Decimal = unspecified, int(math.Max(xxx, yyy))
if a.Flen == types.UnspecifiedLength || b.Flen == types.UnspecifiedLength {
    return
}
....
```

File: expression/builtin_arithmetic.go [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+// setFlenDecimal4Int is called to set proper `Flen` and `Decimal` of return
+// type according to the two input parameter's types.
+func (c *arithmeticPlusFunctionClass) setFlenDecimal4Int(retTp, a, b *types.FieldType) {
+	retTp.Decimal = 0
+	if a.Flen == types.UnspecifiedLength || b.Flen == types.UnspecifiedLength {
+		retTp.Flen = mysql.MaxIntWidth
+	} else {
+		retTp.Flen = int(math.Max(float64(a.Flen), float64(b.Flen))) + 1
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: I think we can set retTp to mysql.MaxIntWidth,
the length of int is not that important,
set it to the default value make the code more readable.

File: expression/builtin_arithmetic.go [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+		return nil, errors.Trace(err)
+	}
+	tpA, tpB := args[0].GetType(), args[1].GetType()
+	tcA, tcB := numericContextResultType(tpA), numericContextResultType(tpB)
+	if tcA == types.ClassReal || tcB == types.ClassReal {
+		bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpReal, tpReal, tpReal)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if mysql.HasUnsignedFlag(tpA.Flag) || mysql.HasUnsignedFlag(tpB.Flag) {
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: we set unsigned flag here?

File: expression/builtin_arithmetic.go (hunk 10 above)
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: no, `types.UnspecifiedLength` is -1, we can not simply use math.Max here. when a parameter's decimal is unspecified, the result's decimal should be unspecified as well.

File: expression/builtin_arithmetic.go (hunk 12 above)
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: no, result type of function plus with two float point parameters is always unsigned

File: expression/builtin_arithmetic.go [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+var (
+	_ builtinFunc = &builtinArithmeticPlusRealSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntUnsignedSig{}
+	_ builtinFunc = &builtinArithmeticPlusDecimalSig{}
+	_ builtinFunc = &builtinArithmeticSig{}
+)
+
+func numericContextResultType(ft *types.FieldType) types.TypeClass {
```
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: done

File: expression/builtin_arithmetic.go (hunk 11 above)
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: ok

File: expression/builtin_arithmetic_test.go (hunk 8 above)
Comment (zz-jason) [inline expression/builtin_arithmetic_test.go]: done

File: util/types/etc.go [hunk 14]
```diff
@@ -149,3 +149,13 @@ func InvOp2(x, y interface{}, o opcode.Op) (interface{}, error) {
 func overflow(v interface{}, tp byte) error {
 	return ErrOverflow.Gen("constant %v overflows %s", v, TypeStr(tp))
 }
+
+// IsTemporalType checks if a type is a temporal type.
+func IsTemporalType(tp byte) bool {
```
Comment (XuHuaiyu) [inline util/types/etc.go]: s/ IsTemporalType/ IsTypeTemporal

File: expression/builtin_arithmetic.go [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+	retTp.Flen = mysql.MaxIntWidth
+}
+
+// setFlenDecimal4Real is called to set proper `Flen` and `Decimal` of return
+// type according to the two input parameter's types.
+func (c *arithmeticPlusFunctionClass) setFlenDecimal4RealOrDecimal(retTp, a, b *types.FieldType, isReal bool) {
+	retTp.Decimal = int(math.Max(float64(a.Decimal), float64(b.Decimal)))
+	retTp.Flen = types.UnspecifiedLength
+	if a.Decimal != types.UnspecifiedLength && b.Decimal != types.UnspecifiedLength {
+		retTp.Decimal = int(math.Max(float64(a.Decimal), float64(b.Decimal)))
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: why we need this line since there is line 69.

File: expression/builtin_arithmetic.go [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+		if a.Flen == types.UnspecifiedLength || b.Flen == types.UnspecifiedLength {
+			retTp.Flen = types.UnspecifiedLength
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: ditto

File: expression/builtin_arithmetic.go (hunk 15 above)
Comment (zz-jason) [inline expression/builtin_arithmetic.go]: line 69~70 should be removed.

File: expression/builtin_arithmetic.go [hunk 17]
```diff
@@ -66,8 +66,6 @@ func (c *arithmeticPlusFunctionClass) setFlenDecimal4Int(retTp, a, b *types.Fiel
 // setFlenDecimal4Real is called to set proper `Flen` and `Decimal` of return
 // type according to the two input parameter's types.
 func (c *arithmeticPlusFunctionClass) setFlenDecimal4RealOrDecimal(retTp, a, b *types.FieldType, isReal bool) {
-	retTp.Decimal = int(math.Max(float64(a.Decimal), float64(b.Decimal)))
-	retTp.Flen = types.UnspecifiedLength
 	if a.Decimal != types.UnspecifiedLength && b.Decimal != types.UnspecifiedLength {
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: if a.Decimal **==** types.UnspecifiedLength **||** b.Decimal **==** types.UnspecifiedLength

we set flen to unspecified?

Comment (zz-jason) [inline expression/builtin_arithmetic.go]: yes

**Comment by XuHuaiyu** [comment]
LGTM 
PTAL @winkyao

File: expression/builtin_arithmetic.go [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+
+var (
+	_ functionClass = &arithmeticPlusFunctionClass{}
+	_ functionClass = &arithmeticFunctionClass{}
+)
+
+var (
+	_ builtinFunc = &builtinArithmeticPlusRealSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntUnsignedSig{}
```
Comment (XuHuaiyu) [inline expression/builtin_arithmetic.go]: never been used,
add test for this

Comment (zz-jason) [inline expression/builtin_arithmetic.go]: a mistake, it should be used when plus two unsigned int values

**Comment by XuHuaiyu** [comment]
PTAL @winkyao

File: expression/builtin_arithmetic.go [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+	sc := s.ctx.GetSessionVars().StmtCtx
+	a, isNull, err := s.args[0].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	b, isNull, err := s.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	if (a > 0 && b > math.MaxInt64-a) || (a < 0 && b < 0 && a < math.MinInt64-b) {
```
Comment (zimulala) [inline expression/builtin_arithmetic.go]: Can we remove the condition of `a > 0` and `a < 0 && b < 0 `?

File: expression/builtin_arithmetic.go [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+	sc := s.ctx.GetSessionVars().StmtCtx
+	a, isNull, err := s.args[0].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	if (a > 0 && b > math.MaxFloat64-a) || (a < 0 && b < 0 && a < -math.MaxFloat64-b) {
```
Comment (zimulala) [inline expression/builtin_arithmetic.go]: Ditto.

File: expression/builtin_arithmetic.go (hunk 19 above)
Comment (zimulala) [inline expression/builtin_arithmetic.go]: Gotcha.

**Comment by zz-jason** [comment]
@zimulala updated, PTAL again.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**